rfcs/pg-function-improvements.md

```diff
@@ -0,0 +1,45 @@
+## Improvements to GraphQL API over SQL functions
+
+Currently we only support stable/immutable functions which return a list of rows.
+The rationale behind this decision was to avoid building a permission system
+for functions. The permissions are implicitly inferred as follows:
+
+> A function is allowed for a role only if the table it returns is allowed.
```

Comment:
I've just finished getting acquainted with the details of postgres functions, as well as hasura's docs on computed fields and function tracking; I haven't gotten into the code just yet, so this may be a dumb question...

" if the table it returns ... "

I think what you mean precisely here is: "if it returns a composite type implicitly created during `CREATE TABLE` on table..."

Not trying to be pedantic, this was actually a little confusing to me

Reply:
I think there's also an interesting question here:

should it be that we attach permissions, not to tables, _but to types_?

If so then we could allow, e.g. someone to `CREATE TYPE` and then define permissions on it and then create and track a function that returns that type (maybe this is already how things work and I just forget...)

Reply:
> Not trying to be pedantic, this was actually a little confusing to me. I think I could improve our docs a little here

:+1: 

> should it be that we attach permissions, not to tables, but to types?

This is something that has been on my mind for quite sometime now. It definitely makes sense to attach 'select' permissions to Postgres's composite types but we'll need to figure out some things:
1. `limit` in select permissions? Does it make sense to attach it to a type? Or should we introduce a notion of 'resolvers' and `limit` is attached to a resolver?
1. What about other permissions: insert/update/delete? Do the features that we have currently extend well for types?
1. What is the user experience gonna look like? i.e, currently the building blocks for user's GraphQL schema are mostly tracking tables, and defining relationships, both of which are built around the notion of 'tables/views'.

Reply:
FWIW, I think we should see if there is an incremental path from table types (which is what our users have asked for mainly) to arbitrary postgres composite types for functions and scope this only to table types if possible.